### PR TITLE
Get referral acquisition history

### PR DESCRIPTION
**Pull request by kfedor**

### CODE REVIEW

File: users/users_referrals.go [hunk 1]
```diff
@@ -35,3 +36,33 @@ WHERE referrals.referred_by = :user_id ORDER BY from_agenda DESC, referrals.crea
...
+
+	var results []struct {
+		Count uint64
+	}
+
+	if err := u.db.PrepareExecuteTyped(sql, params, &results); err != nil {
+		return 0, errors.Wrap(err, "failed to get referred_by user count")
+	}
+
+	return results[0].Count, nil
```
[the remark this answers is not in the document]
Comment (kfedor) [reply users/users_referrals.go]: it was initial version

File: users/users_referrals.go [hunk 2]
```diff
@@ -35,3 +36,90 @@ WHERE referrals.referred_by = :user_id ORDER BY from_agenda DESC, referrals.crea
...
 	return queryResult, nil
 }
+
+func (u *users) GetReferralAcquisitionHistory(ctx context.Context, id UserID, days uint64) ([]*ReferralAcquisition, error) {
+	if ctx.Err() != nil {
+		return nil, errors.Wrap(ctx.Err(), "failed to get acquisition history because context failed")
+	}
+
+	nSecToDays := time.Hour.Nanoseconds() * 24 //nolint:gomnd //Nanoseconds in day
+	today := time.Now().UTC().Unix() / 86400   //nolint:gomnd //This is number of seconds in day
```
Comment (azayats-leftclick) [inline users/users_referrals.go:46]: candidate to const?

Comment (kfedor) [reply users/users_referrals.go:46]: function call is always not a const in go

File: users/users_referrals.go [hunk 3]
```diff
@@ -35,3 +36,90 @@ WHERE referrals.referred_by = :user_id ORDER BY from_agenda DESC, referrals.crea
...
+		"AND -1*(created_at/:nsec_to_days-:today) IN (%v) "+
+		"GROUP BY (created_at/:nsec_to_days-:today) ORDER BY created_at DESC", sql1, daysRange)
+
+	params := map[string]interface{}{
+		"user_id":      id,
+		"nsec_to_days": nSecToDays,
+		"today":        today,
+	}
+
+	var results []referralAcquisition
```
Comment (victor-oancea-leftclick-io) [inline users/users_referrals.go:106]: array of pointers, always, performance is better for unmarshslling

File: users/users_referrals.go [hunk 4]
```diff
@@ -35,3 +36,90 @@ WHERE referrals.referred_by = :user_id ORDER BY from_agenda DESC, referrals.crea
...
+	}
+
+	return results, nil
+}
+
+func (u *users) getT2Stats(id UserID, nSecToDays, today int64, daysRange string) ([]referralAcquisition, error) {
+	sql1 := fmt.Sprintf("SELECT id FROM USERS WHERE REFERRED_BY = :user_id AND -1*(created_at/:nsec_to_days-:today) IN (%v)", daysRange)
+
+	sql := fmt.Sprintf("SELECT COUNT(*), created_at FROM users WHERE referred_by IN (%v) "+
+		"AND -1*(created_at/:nsec_to_days-:today) IN (%v) "+
```
Comment (victor-oancea-leftclick-io) [inline users/users_referrals.go:97]: created_at >= :pastNanos and created_at <= :nowNanos

the same for all sqls

File: users/users_referrals.go [hunk 5]
```diff
@@ -35,3 +36,90 @@ WHERE referrals.referred_by = :user_id ORDER BY from_agenda DESC, referrals.crea
...
+		return nil, errors.Wrap(err, "failed to get referred_by user count")
+	}
+
+	return results, nil
+}
+
+func (u *users) getT2Stats(id UserID, nSecToDays, today int64, daysRange string) ([]referralAcquisition, error) {
+	sql1 := fmt.Sprintf("SELECT id FROM USERS WHERE REFERRED_BY = :user_id AND -1*(created_at/:nsec_to_days-:today) IN (%v)", daysRange)
+
+	sql := fmt.Sprintf("SELECT COUNT(*), created_at FROM users WHERE referred_by IN (%v) "+
```
Comment (victor-oancea-leftclick-io) [inline users/users_referrals.go:96]: the column you return is the same one you group by

File: users/users_referrals.go [hunk 6]
```diff
@@ -35,3 +36,90 @@ WHERE referrals.referred_by = :user_id ORDER BY from_agenda DESC, referrals.crea
...
+		"AND -1*(created_at/:nsec_to_days-:today) IN (%v) "+
+		"GROUP BY (created_at/:nsec_to_days-:today) ORDER BY created_at DESC", sql1, daysRange)
```
Comment (victor-oancea-leftclick-io) [inline users/users_referrals.go:98]: you could do (created_at - :pastNanos)/ 8.64E13, it would give you 1,2,3,4,5

File: users/users_referrals.go [hunk 7]
```diff
@@ -35,3 +36,90 @@ WHERE referrals.referred_by = :user_id ORDER BY from_agenda DESC, referrals.crea
...
+	var results []referralAcquisition
+
+	if err := u.db.PrepareExecuteTyped(sql, params, &results); err != nil {
+		return nil, errors.Wrap(err, "failed to get referred_by user count")
+	}
+
+	return results, nil
+}
+
+func (u *users) getT2Stats(id UserID, nSecToDays, today int64, daysRange string) ([]referralAcquisition, error) {
```
Comment (victor-oancea-leftclick-io) [inline users/users_referrals.go:93]: lets have 1 big sql for both, and try not to use union.

Also, pls lets stop breaking the sql into small pieces, lets have a single var big sql without any %v

**Review by victor-oancea-leftclick-io — CHANGES_REQUESTED** [review]

File: users/users_referrals.go (hunk 4 above)
Comment (kfedor) [reply users/users_referrals.go:97]: mb "between :pastnanos and :nownanos" ?
btw using such case I can't query several days in single query

File: users/users_referrals.go (hunk 7 above)
Comment (kfedor) [reply users/users_referrals.go:93]: ok i'll try
but why one SQL with two queries is better then two sqls ? db will execute two.
two sqls is easier to read and fix

File: users/users_referrals.go (hunk 3 above)
Comment (kfedor) [reply users/users_referrals.go:106]: done

File: users/users_referrals.go (hunk 4 above)
Comment (victor-oancea-leftclick-io) [reply users/users_referrals.go:97]: You check timestamps in a range

Comment (victor-oancea-leftclick-io) [reply users/users_referrals.go:97]: [{days}*8,64E13, nowTS]

File: users/users_referrals.go (hunk 5 above)
Comment (victor-oancea-leftclick-io) [reply users/users_referrals.go:96]: Its gonna be a different value

File: users/users_referrals.go (hunk 7 above)
Comment (victor-oancea-leftclick-io) [reply users/users_referrals.go:93]: Cuz of the RTT

File: users/users_referrals.go (hunk 5 above)
Comment (kfedor) [reply users/users_referrals.go:96]: why? I group by days gone and output by days gone

Comment (victor-oancea-leftclick-io) [reply users/users_referrals.go:96]: You need to group by the date, and return the date . 2022-05-19

File: users/users_referrals.go (hunk 6 above)
Comment (kfedor) [reply users/users_referrals.go:98]: sql := fmt.Sprintf("SELECT COUNT(*), created_at FROM users WHERE referred_by IN "+
		"(SELECT id FROM USERS WHERE REFERRED_BY = :user_id AND -1*((created_at-:today)/:nsec_to_days) IN (%v)) "+
		"AND -1*((created_at-:today)/:nsec_to_days) IN (%v) "+
		"GROUP BY (created_at-:today)/:nsec_to_days ORDER BY created_at DESC", daysRange, daysRange)

where :today is time.Now().UTC().UnixNano()
nsec_to_days is 8.64E13
right?

File: users/users_referrals.go (hunk 7 above)
Comment (kfedor) [reply users/users_referrals.go:93]: how to do this?
SELECT ( SELECT COUNT(*) ), ( SELECT COUNT(*) ) FROM 
is working
but 
SELECT ( SELECT COUNT(*), created_at ), ( SELECT COUNT(*), created_at ) FROM 
fails
are there good SQL clients for tarantool ?

File: users/users_referrals.go (hunk 6 above)
Comment (victor-oancea-leftclick-io) [reply users/users_referrals.go:98]: Forget about it, ill do the sql, fix the other comments

File: users/users_referrals.go [hunk 8]
```diff
@@ -35,3 +36,90 @@ WHERE referrals.referred_by = :user_id ORDER BY from_agenda DESC, referrals.crea
...
+		tier + 1                                              AS tier
+	FROM referrals
+	JOIN users i
+	ON referrals.id = i.referred_by
+	AND i.created_at >= :pastNanos
+	AND i.created_at <= :nowNanos
+	WHERE tier < 3
+	)
+	SELECT
+		CAST(TOTAL(referrals.t1) AS INT),
```
Comment (victor-oancea-leftclick-io) [inline users/users_referrals.go:82]: put the aliases back

**Review by victor-oancea-leftclick-io — CHANGES_REQUESTED** [review]

File: users/users_referrals.go (hunk 8 above)
Comment (kfedor) [reply users/users_referrals.go]: done, why do they matter?

Comment (victor-oancea-leftclick-io) [reply users/users_referrals.go]: for the same reason naming funcs, methods, vars, args properly matters, so ppl would better understand whats going on

**Review by nusanovkornilov-leftclick — DISMISSED** [review]

**Review by nusanovkornilov-leftclick — APPROVED** [review]

**Pull request merged**